pkg/controller/nodes/dynamic/handler.go

```diff
@@ -117,31 +118,39 @@ func (d dynamicNodeTaskNodeHandler) handleDynamicSubNodes(ctx context.Context, n
 		})
 
 		if err != nil {
-			return handler.UnknownTransition, err
+			return handler.UnknownTransition, prevState, err
 		}
 
 		if ee != nil {
 			if ee.IsRecoverable {
-				return trns.WithInfo(handler.PhaseInfoRetryableFailureErr(ee.ExecutionError, trns.Info().GetInfo())), nil
+				return trns.WithInfo(handler.PhaseInfoRetryableFailureErr(ee.ExecutionError, trns.Info().GetInfo())), handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing}, nil
 			}
-			return trns.WithInfo(handler.PhaseInfoFailureErr(ee.ExecutionError, trns.Info().GetInfo())), nil
+
+			return trns.WithInfo(handler.PhaseInfoFailureErr(ee.ExecutionError, trns.Info().GetInfo())), handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing}, nil
 		}
 	}
 
-	return trns, nil
+	return trns, newState, nil
 }
 
 func (d dynamicNodeTaskNodeHandler) Handle(ctx context.Context, nCtx handler.NodeExecutionContext) (handler.Transition, error) {
 	ds := nCtx.NodeStateReader().GetDynamicNodeState()
 	var err error
 	var trns handler.Transition
 	newState := ds
+	logger.Infof(ctx, "Dynamic handler.Handle's called with phase %v.", ds.Phase)
 	if ds.Phase == v1alpha1.DynamicNodePhaseExecuting {
-		trns, err = d.handleDynamicSubNodes(ctx, nCtx)
+		trns, newState, err = d.handleDynamicSubNodes(ctx, nCtx, ds)
 		if err != nil {
 			logger.Errorf(ctx, "handling dynamic subnodes failed with error: %s", err.Error())
 			return trns, err
 		}
+	} else if ds.Phase == v1alpha1.DynamicNodePhaseFailing {
```

Comment:
I think this is an excellent addition, but we should probably abort in this case. So Dynamic now has 3 potential steps
New -> Executing (potentially -> Failing in case of failures)
then the abort for the node could propagate abort, if the state is `failing/executing`

Reply:
and ofcourse finalize will be simplified